zebrad/tests/common/lightwalletd/sync.rs

```diff
@@ -0,0 +1,206 @@
+        );
+        while !is_lightwalletd_finished.load(Ordering::SeqCst) {
+            // Just keep checking the Zebra logs for errors...
+            if wait_for_zebrad_tip {
+                // Make sure the sync is still finished, this is logged every minute or so.
+                zebrad_mut.expect_stdout_line_matches(crate::common::sync::SYNC_FINISHED_REGEX)?;
+            } else {
+                // Use sync progress logs, which are logged every minute or so.
+                zebrad_mut.expect_stdout_line_matches(crate::common::sync::SYNC_PROGRESS_REGEX)?;
+            }
+        }
+
+        Ok(zebrad_mut)
+    };
+
+    // Wait until `lightwalletd` has synced to Zebra's tip.
+    // Calls `lightwalletd`'s gRPCs and Zebra's JSON-RPCs.
+    // Also checks `lightwalletd`'s logs for errors.
+    //
+    // `zebrad` syncs can take a long time,
+    // so we need to check that `lightwalletd` has synced to the tip in parallel.
+    let lightwalletd_mut = &mut lightwalletd;
+    let lightwalletd_wait_fn = || -> Result<_> {
+        tracing::info!(
+            ?test_type,
+            "lightwalletd is waiting for zebrad to sync to the tip..."
+        );
+        while !is_zebrad_finished.load(Ordering::SeqCst) {
+            // Just keep checking the `lightwalletd` logs for errors.
+            // It usually logs something every 30-90 seconds,
+            // but there's no specific message we need to wait for here.
+            assert!(
+                lightwalletd_mut.wait_for_stdout_line(None),
+                "lightwalletd output unexpectedly finished early",
+            );
+        }
+
+        tracing::info!(?test_type, "waiting for lightwalletd to sync to the tip...");
+        while !are_zebrad_and_lightwalletd_tips_synced(lightwalletd_rpc_port, zebra_rpc_address)? {
+            let previous_check = Instant::now();
+
+            // To improve performance, only check the tips occasionally
+            while previous_check.elapsed() < TIP_CHECK_RATE_LIMIT {
+                assert!(
+                    lightwalletd_mut.wait_for_stdout_line(None),
+                    "lightwalletd output unexpectedly finished early",
+                );
+            }
+        }
+
+        // Tell the other thread that `lightwalletd` has finished
+        is_lightwalletd_finished.store(true, Ordering::SeqCst);
+
+        Ok(lightwalletd_mut)
+    };
+
+    // Run both threads in parallel, automatically propagating any panics to this thread.
+    std::thread::scope(|s| {
+        // Launch the sync-waiting threads
+        let zebrad_thread = s.spawn(|| {
+            zebrad_wait_fn().expect("test failed while waiting for zebrad to sync");
+        });
+
+        let lightwalletd_thread = s.spawn(|| {
+            let lightwalletd_result = lightwalletd_wait_fn();
+            is_lightwalletd_finished.store(true, Ordering::SeqCst);
+            lightwalletd_result.expect("test failed while waiting for lightwalletd to sync");
```

Comment:
It looks like `is_lightwalletd_finished` is being set to true after this thread is joined. It could also update the finished markers with a panic hook.

```suggestion
            lightwalletd_wait_fn().expect("test failed while waiting for lightwalletd to sync.");
```

Reply:
Thanks, I meant to do that cleanup, but forgot about it.

> It could also update the finished markers with a panic hook.

That doesn't work, because `&mut zebra_test::command::TestChild<TempDir>` is not unwind-safe.
See the TODO note below.